src/controller.rs

```diff
@@ -185,26 +185,124 @@ fn apply_mappings(
+                    .data(subtree);
+                source.metadata.annotations = source_metadata.annotations;
+                source.metadata.labels = source_metadata.labels;
+                template = clone_resource(&source, target_ref, &target_namespace, &ar)?;
+            }
+            Mapping {
+                from_field_path: _,
+                to_field_path: Some(to_field_path),
+            } => {
+                if to_field_path.starts_with("metadata.") {
+                    // DynamicObject's metadata is not a serde_json::value::Value,
+                    // but we can convert to/from that and use the same code to update
+                    // it as we do for spec/status in .data
+                    let mut metadata = serde_json::value::to_value(template.metadata.clone())?;
+                    set_field_path(
+                        &mut metadata,
+                        &to_field_path.strip_prefix("metadata.").unwrap(),
+                        subtree.clone(),
+                    )?;
+                    template.metadata = serde_json::value::from_value(metadata)?;
+                } else {
+                    set_field_path(&mut template.data, &to_field_path, subtree.clone())?;
+                }
+            }
         }
         let dbg = serde_json::to_string_pretty(&template)?;
         debug!(%dbg, "after");
     }
     Ok(template)
 }
 
+// extract metadata that we can set on a DynamicObject from a serde_json value subtree.
+fn convert_metadata(subtree: &serde_json::Value) -> ObjectMeta {
+    let mut metadata = ObjectMeta {
+        ..Default::default()
+    };
+    if let serde_json::Value::Object(map) = &subtree["annotations"] {
+        metadata.annotations = Some(
+            map.iter()
+                .map(|(k, v)| (k.to_string(), v.as_str().unwrap().to_string()))
+                .collect::<BTreeMap<String, String>>(),
+        );
+    }
+    if let serde_json::Value::Object(map) = &subtree["labels"] {
+        metadata.labels = Some(
+            map.iter()
+                .map(|(k, v)| (k.to_string(), v.as_str().unwrap().to_string()))
+                .collect::<BTreeMap<String, String>>(),
+        );
+    }
+    metadata
+}
+
+// extract GVKN from a k8s resource in an arbitrary serde_json subtree.
+fn get_ar_from_subtree(subtree: &serde_json::Value) -> Result<ApiResource> {
+    let api_version = subtree["apiVersion"]
+        .as_str()
+        .ok_or(Error::MalformedInnerResource(
+            "failed to parse apiVersion".to_string(),
+        ))?;
+    // account for group-less resources by extracting version from the end first,
+    // then group if there is a term remaining.
+    let mut gv_terms = api_version.split("/").collect::<Vec<_>>();
+    let version = gv_terms
+        .pop()
+        .ok_or(Error::MalformedInnerResource(
+            "failed to parse apiVersion".to_string(),
+        ))?
+        .to_string();
+    let group = gv_terms.pop().unwrap_or("").to_string();
```

Comment:
fun fact: `split` can return a `DoubleEndedIterator`, if you split on a `char` but not on a `str`.

Thus:

```suggestion
    let mut gv_terms = api_version.split('/').rev();
    let version = gv_terms
        .next()
        .ok_or(Error::MalformedInnerResource(
            "failed to parse apiVersion".to_string(),
        ))?
        .to_string();
    let group = gv_terms.next().unwrap_or("").to_string();
```

Reply:
(not important, just a rabbit hole I recently went through in another project)

Reply:
urgh this is what i tried to do first but ran into the "it's not a double ended iterator" message. thanks!